neomutt: add IMAP support

Cybolic opened the pull request:
### Description

This adds support for IMAP accounts in NeoMutt as well as IMAP or maildir overrides for individual mailboxes.
Fixes #4376.

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

<!--
If you are updating a module, please @ people who are in its `meta.maintainers` list.
If in doubt, check `git blame` for whoever last touched something.
-->
@jvanbruegge @rycee @berbiche @michaeladler @sumnerevans

Cybolic commented:
Currently, this is not backwards compatible, as it defaults to using IMAP for a mailbox whenever possible.
Keeping in mind that NeoMutt can use a mix of maildir and IMAP mailboxes, I'm planning to add opt-in support for IMAP accounts and mailboxes through the following options:

1. `accounts.email.accounts.<name>.neomutt.mailboxType = types.enum [ "maildir" "imap" ]`
Use either `IMAP` or `maildir` as the default for all mailboxes of this account (it would still default to `maildir` to keep compatibility). If no `extraMailBoxes` are set, this would only hardcode the mailbox type for `inbox`.
  
2. `accounts.email.accounts.<name>.neomutt.extraMailBoxes = listOf { mailbox = types.string; name = types.nullOr types.string; type = types.enum [ "maildir" "imap" ]}` (i.e. add a new `type` option to the `extraMailBoxes` type)
Override the default mailbox type for an individual mailbox.

Feedback on this approach is very welcome.

chayleaf commented on modules/programs/neomutt.nix:
```diff
@@ -145,29 +218,40 @@ let
-        if isString extra then
-          ''mailboxes "${account.maildir.absPath}/${extra}"''
+        let
+          mailboxroot = if isString extra == false && extra.type == "imap" then
```
see my comment on line 13

chayleaf commented on modules/programs/neomutt.nix:
```diff
@@ -145,29 +218,40 @@ let
+        let
+          mailboxroot = if isString extra == false && extra.type == "imap" then
+            accountRootIMAP account
+          else if isString extra == false && extra.type == "maildir" then
```
same as above

chayleaf reviewed (commented):
other than the small nits, I've checked that it works and the old behavior does get preserved without `neomutt.mailboxType = "imap";`.

Cybolic commented on modules/programs/neomutt.nix:
```diff
@@ -101,6 +154,26 @@ let
+        ":${toString account.imap.port}"
+      else
+        "";
+      protocol = if account.imap.tls.enable || account.imap.port == 993 then
```
No, it isn't done anywhere else in home-manager, so you're probably right that this should be left out.

Cybolic commented:
Thanks @chayleaf for those suggestions! This is the first time I use Nix/HM for anything, so I really do appreciate those tweaks.
I've implemented all your suggestions and I do agree that they simplify things.

I've added an extra change since the initial PR, which is to enable `ssl_force_tls` based on the value of `imap.tls` instead of `smtp.tls` since that's the type of connection NeoMutt actually uses the setting for. The tests still pass after this change.

chayleaf reviewed (approved):
I'm not a HM contributor, but LGTM! Please note that in nixpkgs commits are typically squashed, and I recommend doing it here to a reasonable extent too (i.e. adding tests in a separate commit is fine, but any fixes should probably be applied to the original commit rather than added in a new commit).

Cybolic commented:
I just fixed the failing test. Also, just so Github can pick it up, this PR would fix #4376.
